aa_stripe/models.py

```diff
@@ -58,6 +59,89 @@ class Meta:
+        max_length=3, default="usd", help_text=_("If amount_off has been set, the three-letter ISO code for the"
+                                                 "currency of the amount to take off."))
+    duration = models.CharField(
+        max_length=255, choices=DURATION_CHOICES,
+        help_text=_("Describes how long a customer who applies this coupon will get the discount."))
+    duration_in_months = models.PositiveIntegerField(
+        blank=True, null=True, help_text=_("If duration is repeating, the number of months the coupon applies."
+                                           "Null if coupon duration is forever or once."))
+    livemode = models.BooleanField(
+        default=False, help_text=_("Flag indicating whether the object exists in live mode or test mode."))
+    max_redemptions = models.PositiveIntegerField(
+        blank=True, null=True,
+        help_text=_("Maximum number of times this coupon can be redeemed, in total, before it is no longer valid."))
+    metadata = JSONField(help_text=_("Set of key/value pairs that you can attach to an object. It can be useful for"
+                                     "storing additional information about the object in a structured format."))
+    percent_off = models.PositiveIntegerField(
+        blank=True, null=True,
+        help_text=_("Percent that will be taken off the subtotal of any invoicesfor this customer for the duration of"
+                    "the coupon. For example, a coupon with percent_off of 50 will make a $100 invoice $50 instead."))
+    redeem_by = models.DateTimeField(
+        blank=True, null=True, help_text=_("Date after which the coupon can no longer be redeemed."))
+    times_redeemed = models.PositiveIntegerField(
+        default=0, help_text=_("Number of times this coupon has been applied to a customer."))
+    valid = models.BooleanField(
+        default=False,
+        help_text=_("Taking account of the above properties, whether this coupon can still be applied to a customer."))
+    created = models.DateTimeField()
+    is_deleted = models.BooleanField(default=False)
+
+    def save(self, *args, **kwargs):
+        stripe.api_key = settings.STRIPE_API_KEY
+        if self.pk:
+            try:
+                coupon = stripe.Coupon.retrieve(self.coupon_id)
+                coupon.metadata = self.metadata
+                coupon.save()
+
+                # update all fields in the local object in case someone tried to change them
+                self.coupon_id = coupon.id
+                readonly_fields = [
+                    "amount_off", "currency", "duration", "duration_in_months", "livemode", "max_redemptions",
+                    "percent_off", "redeem_by", "times_redeemed", "valid"
+                ]
+                for field in readonly_fields:
+                    setattr(self, field, getattr(coupon, field))
+            except stripe.error.InvalidRequestError:
+                self.is_deleted = True
```

Comment:
I added validation in admin form (I thought about just catching Stripe API exceptions in form's save() method, but for me it looked a bit hacky, I decided to go for form validation, as it seems more reasonable).